elki-outlier/src/main/java/elki/outlier/density/HySortOD.java

```diff
@@ -0,0 +1,694 @@
+   */
+  private DensityStrategy strategy;
+
+  /**
+   * Constructor with parameters.
+   * 
+   * @param b Number of bins
+   * @param minSplit Threshold to balance the tree strategy
+   */
+  public HySortOD(int b, int minSplit) {
+    super();
+    this.b = b;
+    this.l = 1 / (double) this.b;
+    this.minSplit = minSplit;
+    if(this.minSplit > 0) {
+      this.strategy = new TreeStrategy(this.minSplit);
+    }
+    else {
+      this.strategy = new NaiveStrategy();
+    }
+  }
+
+  public OutlierResult run(Database db, Relation<V> relation) {
+
+    // Output data storage
+    WritableDoubleDataStore scores = DataStoreUtil.makeDoubleStorage(relation.getDBIDs(), DataStoreFactory.HINT_DB);
+    StepProgress stepprog = LOG.isVerbose() ? new StepProgress(3) : null;
+
+    LOG.beginStep(stepprog, 1, "Ordering of hypercubes lexicographically according to their coordinates.");
+    final List<Hypercube> H = getSortedHypercubes(relation);
+
+    LOG.beginStep(stepprog, 2, "Obtaining densities per hypercube.");
+    final int[] W = this.strategy.buildIndex(H).getDensities();
+
+    // Track minimum and maximum scores
+    DoubleMinMax minmax = new DoubleMinMax();
+    {
+      // compute score by hypercube
+      LOG.beginStep(stepprog, 3, "Computing hypercube scores");
+
+      FiniteProgress prog = LOG.isVerbose() ? new FiniteProgress("HySortOD scores", relation.size(), LOG) : null;
+      DBIDArrayIter iter = (DBIDArrayIter) relation.iterDBIDs();
+      for(int hypercube = 0; hypercube < H.size(); hypercube++) {
+        final IntegerArray instances = H.get(hypercube).getInstances();
+        final double hypercubeScore = score(W[hypercube]);
+        minmax.put(hypercubeScore);
+        for(int instance = 0; instance < instances.size; instance++) {
+          final int index = instances.get(instance);
+          scores.putDouble(iter.seek(index), hypercubeScore);
+          LOG.incrementProcessed(prog);
+        }
+      }
+      LOG.ensureCompleted(prog);
+    }
+
+    // Wrap the result in the standard containers
+    OutlierScoreMeta meta = new BasicOutlierScoreMeta(
+        // Actually observed minimum and maximum values
+        minmax.getMin(), minmax.getMax(),
+        // Theoretical minimum and maximum: no variance to infinite variance
+        0, Double.POSITIVE_INFINITY);
+    DoubleRelation rel = new MaterializedDoubleRelation("HySortOD", relation.getDBIDs(), scores);
+    return new OutlierResult(meta, rel);
+  }
+
+  /**
+   * Create and sort hypercubes considering their coordinates.
+   * 
+   * @param relation Data to process
+   * @return Hypercubes sorted
+   */
+  private List<Hypercube> getSortedHypercubes(Relation<V> relation) {
+    TreeSet<Hypercube> sorted = new TreeSet<HySortOD<V>.Hypercube>();
+
+    // Iterate over all objects
+    for(DBIDArrayIter iter = (DBIDArrayIter) relation.iterDBIDs(); iter.valid(); iter.advance()) {
+      Hypercube h = new Hypercube(relation.get(iter).toArray(), this.l);
```

Comment:
This creates a copy of each vector, then a hypercube object for it, a fairly expensive operation.
Likely the algorithm could be made much faster if the lookup would not require object creation.